src/cert/ssl_cert_provider.cpp

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (C) 2018 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "ssl_cert_provider.h"
+
+#include <openssl/bio.h>
+#include <openssl/pem.h>
+#include <openssl/rsa.h>
+#include <openssl/x509.h>
+
+#include <memory>
+#include <vector>
+
+namespace mp = multipass;
+
+namespace
+{
+struct EVPKeyDeleter
+{
+    void operator()(EVP_PKEY* key)
+    {
+        EVP_PKEY_free(key);
+    }
+};
+
+struct RSADeleter
+{
+    void operator()(RSA* rsa)
+    {
+        RSA_free(rsa);
+    }
+};
+
+struct BIGNUMDeleter
+{
+    void operator()(BIGNUM* num)
+    {
+        BN_free(num);
+    }
+};
+
+struct X509Deleter
+{
+    void operator()(X509* x509)
+    {
+        X509_free(x509);
+    }
+};
+
+struct BIODeleter
+{
+    void operator()(BIO* bio)
+    {
+        BIO_free(bio);
+    }
+};
+
+auto make_key()
+{
+    std::unique_ptr<EVP_PKEY, EVPKeyDeleter> ec_pkey{EVP_PKEY_new()};
+    if (ec_pkey == nullptr)
+        throw std::runtime_error("Failed to allocate private key structure");
+
+    std::unique_ptr<RSA, RSADeleter> rsa{RSA_new()};
+    if (rsa == nullptr)
+        throw std::runtime_error("Failed to allocate RSA structure");
+
+    std::unique_ptr<BIGNUM, BIGNUMDeleter> value{BN_new()};
+    if (value == nullptr)
+        throw std::runtime_error("Failed to allocate BIGNUM structure");
+
+    if (!BN_set_word(value.get(), RSA_F4))
+        throw std::runtime_error("Failed to set BIGNUM to RSA_F4");
+
+    if (!RSA_generate_key_ex(rsa.get(), 2048, value.get(), nullptr))
```

Comment:
What happens here when we don't have enough entropy?

Reply:
I traced down to see what the source of entropy was...it essentially does a getrandom syscall, without any flags
http://man7.org/linux/man-pages/man2/getrandom.2.html

So potentially it could block only when the entropy pool is not initialized (i.e. /dev/urandom). 

This is similar to OpenSSHKeyProvider.